app/src/main/kotlin/compiler/intermediate/ControlFlowGraphBuilder.kt

```diff
@@ -1,41 +1,42 @@
 package compiler.intermediate
 
-import compiler.utils.referenceHashMapOf
-import compiler.utils.referenceHashSetOf
+import compiler.utils.Ref
+import compiler.utils.mutableRefMapOf
+import compiler.utils.refMapOf
 
 class IncorrectControlFlowGraphError(message: String) : Exception(message)
 
 class ControlFlowGraphBuilder(@JvmField var entryTreeRoot: IFTNode? = null) {
-    private var unconditionalLinks = referenceHashMapOf<IFTNode, IFTNode>()
-    private var conditionalTrueLinks = referenceHashMapOf<IFTNode, IFTNode>()
-    private var conditionalFalseLinks = referenceHashMapOf<IFTNode, IFTNode>()
-    private var treeRoots = referenceHashSetOf<IFTNode>()
+    private var unconditionalLinks = mutableRefMapOf<IFTNode, IFTNode>()
+    private var conditionalTrueLinks = mutableRefMapOf<IFTNode, IFTNode>()
+    private var conditionalFalseLinks = mutableRefMapOf<IFTNode, IFTNode>()
+    private var treeRoots = mutableListOf<Ref<IFTNode>>()
```

Comment:
Any particular reasons for switching from set to list? We do perform lookups on this, so I think that the usage of the set was justified.

Reply:
Isomorphism checking would be quite difficult without this.

Reply:
Basically, the output order of nodes in `ControlFlowGraphBuilder` must be deterministic. I added an optimization to use both a list (for storing order) and a set (for lookups).